### PR TITLE
SecurityPkg: Tcg2Smm: Inspect target address before usage

### PR DESCRIPTION
**Pull request by kuqin12**
## Description

This change uses abstracted interface from MemLib to validate incoming nested pointer before usage to ensure user supplied legitimate NVS buffer for corresponding TCG operations.

- [x] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [ ] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [ ] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [ ] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [ ] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested

This change is validated on proprietary hardware platform.

## Integration Instructions

N/A

### CODE REVIEW

File: SecurityPkg/Tcg/Tcg2Smm/Tcg2Smm.c [hunk 1]
```diff
@@ -86,6 +86,13 @@ TpmNvsCommunciate (
   switch (CommParams->Function) {
     case TpmNvsMmExchangeInfo:
       DEBUG ((DEBUG_VERBOSE, "[%a] - Function requested: MM_EXCHANGE_NVS_INFO\n", __FUNCTION__));
+      // MU_CHANGE TCBZ4378 [BEGIN] - Check for invalid NVS buffer location
+      if (IsBufferOutsideMmValid (CommParams->TargetAddress, sizeof (TCG_NVS))) {
+        DEBUG ((DEBUG_ERROR, "[%a] - NVS buffer in invalid location!\n", __FUNCTION__));
```
Comment (makubacki) [inline SecurityPkg/Tcg/Tcg2Smm/Tcg2Smm.c]: For new code, `__func__` will blend better with the convention currently used in the file in edk2.

```suggestion
        DEBUG ((DEBUG_ERROR, "[%a] - NVS buffer in invalid location!\n", __func__));

```

File: SecurityPkg/Tcg/Tcg2Smm/Tcg2Smm.c
Comment (makubacki) [inline SecurityPkg/Tcg/Tcg2Smm/Tcg2Smm.c:1]: Can't leave a comment near the code due to GitHub UI, but Line 76 is currently using:

```c
  if (!IsCommBufferValid ((UINTN)CommBuffer, TempCommBufferSize)) {
    DEBUG ((DEBUG_ERROR, "[%a] - MM Communication buffer in invalid location!\n", __FUNCTION__));
    return EFI_ACCESS_DENIED;
  }
```

Whereas it has used `IsBufferOutsideMmValid()` [in edk2](https://github.com/tianocore/edk2/blob/master/SecurityPkg/Tcg/Tcg2Smm/Tcg2Smm.c#L76-L79) for a couple years:

```c
  if (!IsBufferOutsideMmValid ((UINTN)CommBuffer, TempCommBufferSize)) {
    DEBUG ((DEBUG_ERROR, "[%a] - MM Communication buffer in invalid location!\n", __func__));
    return EFI_ACCESS_DENIED;
  }
```

File: SecurityPkg/Tcg/Tcg2Smm/Tcg2Smm.c [hunk 2]
```diff
@@ -86,6 +86,13 @@ TpmNvsCommunciate (
   switch (CommParams->Function) {
     case TpmNvsMmExchangeInfo:
       DEBUG ((DEBUG_VERBOSE, "[%a] - Function requested: MM_EXCHANGE_NVS_INFO\n", __FUNCTION__));
+      // MU_CHANGE TCBZ4378 [BEGIN] - Check for invalid NVS buffer location
+      if (IsBufferOutsideMmValid (CommParams->TargetAddress, sizeof (TCG_NVS))) {
```
Comment (makubacki) [inline SecurityPkg/Tcg/Tcg2Smm/Tcg2Smm.c]: Was this intended to be `!IsBufferOutsideMmValid()`?

```suggestion
      if (!IsBufferOutsideMmValid (CommParams->TargetAddress, sizeof (TCG_NVS))) {
```

File: SecurityPkg/Tcg/Tcg2Smm/Tcg2Smm.c
Comment (kuqin12) [reply SecurityPkg/Tcg/Tcg2Smm/Tcg2Smm.c:1]: The edk2 function call will not work if the supervisor copies the comm buffer into SMRAM. In that case, the `IsBufferOutsideMmValid` will fail.

File: SecurityPkg/Tcg/Tcg2Smm/Tcg2Smm.c (hunk 2 above)
Comment (kuqin12) [reply SecurityPkg/Tcg/Tcg2Smm/Tcg2Smm.c]: Nice catch, thanks.

**Review by makubacki — APPROVED** [review]

**Review by apop5 — APPROVED** [review]

**Pull request merged**